doc/developer/design/20240610_unified_compute_introspection.md

```diff
@@ -0,0 +1,134 @@
+We propose to give the compute controller access to the coordinator's `transient_id_gen` by converting it from an `u64` to an `Arc<AtomicU64>` and passing a reference to the compute controller during creation.
+
+The compute controller needs to handle responses for introspection subscribes specially:
+Rather than propagating them to its clients, it needs to translate them into updates of the associated storage-managed collections:
+
+- `Row` updates (both insertions and retractions) are translated into the schema of the storage-managed collection by prepending the replica ID as the first column.
+  The modified `Row`s are then sent to the storage controller for recording under the respective `IntrospectionType`.
+- Subscribe errors and cancellations are translated into deletion requests to the storage controller.
+  A deletion request targets the storage-managed collection identified by the respective `IntrospectionType` and contains a filter on the first column of each contained `Row` that selects the ID of the replica whose introspection subscribe was interrupted.
+
+### Deletion from Storage-managed Collections
+
+The storage controller delegates the writing of storage-managed collections to a background task called the `CollectionManager`.
+The `CollectionManager` accepts `GlobalId`s identifying storage collections and corresponding `(Row, Diff)` updates to be appended to these collections, and appends them at the current system time.
+It does not yet expose a mechanism to delete previously written updates from a collection by reading back the collection contents, determining the necessary retractions, and appending them.
+
+To remove introspection data for dropped/disconnected replicas from the unified introspection relations, we require such a deletion mechanism.
+Note that the compute controller is not itself able to determine the set of necessary retractions.
+While it can be taught to read the contents of storage-managed collections, it doesn't have any way of knowing the time as of which all previously emitted introspection data has been fully written to its target collection.
+The compute controller would risk reading back the target collection's contents too soon and retract only part of the introspection data previously emitted for the disconnected replica.
+
+We propose extending the `CollectionManager`'s API to also accept deletion requests.
+A deletion request contains a collection ID, as well as a `filter` closure `Box<dyn Fn(&Row) -> bool>`.
+The `CollectionManager` handles a deletion request by reading back the target collection at its latest readable time, applying the `filter` closure to each read `Row`, and appending retractions for each `Row` the `filter` returned `true` for.
+Append and deletion requests must be applied by the `CollectionManager` in the same order they have been issued by the client, to guarantee that deletions will retract all previously sent appends.
```

Comment:
I'm assuming you and @aljoscha have already chatted about this, as this has an interesting interaction with https://github.com/MaterializeInc/materialize/pull/27496. I'm +1 on the design you propose in isolation, but I have not yet thought deeply about how it interacts with the 0dt work.

Reply:
#27496 makes this a bit easier because because it makes the `CollectionManager` keep the whole snapshot state in memory, so we can build the retractions from there directly. Eventually we'll still want to move to the design described here, when the `CollectionManager` learns to only keep the `desired`-`persist` diff around.